security/advancedtls/provider.go

```diff
@@ -0,0 +1,184 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package advancedtls
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"google.golang.org/grpc/credentials/tls/certprovider"
+)
+
+const (
+	defaultTestTimeout = 1 * time.Second
+)
+
+// IdentityPemFileProviderOptions contains fields to be filled out by users
+// for obtaining identity credential files.
+type IdentityPemFileProviderOptions struct {
+	CertFile string
+	KeyFile  string
+	Interval time.Duration
+}
+
+// IdentityPemFileProvider keeps identity credential implementations up to
+// date with secrets that they rely on to secure communications on the
+// underlying channel.
+type IdentityPemFileProvider struct {
+	distributor *certprovider.Distributor
+	closeFunc   func()
+}
+
+// RootPemFileProviderOptions contains fields to be filled out by users
+// for obtaining root credential files.
+type RootPemFileProviderOptions struct {
+	TrustFile string
+	Interval  time.Duration
+}
+
+// RootPemFileProvider keeps root credential implementations up to
+// date with secrets that they rely on to secure communications on the
+// underlying channel.
+type RootPemFileProvider struct {
+	distributor *certprovider.Distributor
+	closeFunc   func()
+}
+
+// NewIdentityPemFileProvider uses IdentityPemFileProviderOptions to
+// construct a IdentityPemFileProvider.
+func NewIdentityPemFileProvider(o *IdentityPemFileProviderOptions, quit chan bool) (*IdentityPemFileProvider, error) {
```

Comment:
Can we initiate the quit channel inside the function?

Reply:
The reason I put it as a parameter is because the user could control the channel themselves and close it whenever they want. If we initiate the quit channel inside, is there any way for users to have direct control over it?

Reply:
We have a Close function which will handle the case when they want to close the channel. Do we have any other use cases where we want users to fully control the channel?

Reply:
Oh wait, you are right. I didn't think of that... Will fix :)